src/transform/pre_reduce.rs

```diff
@@ -19,39 +22,125 @@ impl run::AsDef for InertDef {
   }
 }
 
+pub struct PreReduceStats {
+  pub rewrites: Rewrites,
+  pub errors: Vec<String>,
+}
+
 impl Book {
   /// Reduces the definitions in the book individually, except for the skipped
   /// ones.
   ///
   /// Defs that are not in the book are treated as inert defs.
-  pub fn pre_reduce(&mut self, skip: &dyn Fn(&str) -> bool, max_memory: usize, max_rwts: u64) -> Result<(), String> {
+  pub fn pre_reduce(&mut self, skip: &dyn Fn(&str) -> bool, max_memory: usize, max_rwts: u64) -> PreReduceStats {
+    /// State of the pre-reduction algorithm.
+    ///
+    /// Here's how it works:
+    /// - Each definition is visited in reverse topological order (dependencies
+    ///   then dependents)
```

Comment:
I was wrong here; this is just topological order (had the arrows the wrong way).

Would be good IMO to mention that cycles are decided arbitrarily.
```suggestion
    /// - Each definition is visited in topological order (dependencies before
    ///   dependents). In the case of cycles, one will be arbitrarily selected
    ///   to be first.
```